### ISSUE TITLE
Undefined behavior in ElementTransformation::GetIntPoint()

### ISSUE THREAD

**Issue by mlstowell** [mfem/mfem#588]
https://github.com/mfem/mfem/blob/893b127aedb8c495efb63cf1f45d7befe23c4db9/fem/eltrans.hpp#L54

The IntegrationPoint pointer stored in an ElementTransformation object is NULL by default but the GetIntPoint() method dereferences this pointer without checking its value.  I imagine that in practice this pointer is set using SetIntPoint() before any calls to GetIntPoint() are made but can we really be certain of this?

Also, SetIntPoint() could be used to pass in a NULL pointer which would also lead to undefined behavior in GetIntPoint().

I'd like to suggest that we add a method to ElementTransformation class which could be used to check if an IntegrationPoint is set i.e. a method that returns (IntPoint != NULL).  So that nervous developers can call this before calling GetIntPoint().

**Comment by YohannDudouit** [comment]
I personally dislike design that allow attributes to be `NULL`, one of the (many) problem with `NULL` pointers I've tried to explain to @v-dobrev is that we have to always check if `NULL` which has many drawbacks. I think that in this case we shouldn't allow `IntegrationRule` to be `NULL`, if we allow it means that we need to carry all the information to build the `IntegrationPoint` by ourself in some way, but if we now have to carry this information then we don't need this function (which is just bug prone since it can be `NULL` in an unpredictable way).
What I suggest is either this can never be `NULL` (we should _always_ have an `IntegrationRule` associated when building the object), or we remove this method and the associated attribute to avoid bugs. I imagine that for backward compatibility none of this is possible though... Maybe adding a constructor that does not let the user set the `IntegrationRule` to `NULL`, and throwing a warning message saying that the other constructor is deprecated because unsafe, and adding your function `bool HasIntegrationRule() const` is an acceptable path.

**Comment by mlstowell** [comment]
@YohannDudouit , I tend to use a lot of NULL pointers so I cannot say that I agree with you completely.  The trouble is that it is not always possible to have all of the necessary information needed to build all contained objects when the container is being constructed.  With that in mind I tend to use NULL pointers as indicators that the containing object is not yet ready to be used.

However, the trouble with ElementTransformation is more subtle.  The IntegrationPoint class does not require any arguments for its constructor so we could certainly create one inside ElementTransformation by default.  But then we have what looks like a valid integration point even when it really is not valid.  It may have the incorrect dimension and its value is (0,0,0) which may not be what you need.  And furthermore you have no way of knowing if this is the default integration point or one that was intentionally set.  The way the code is written now, calls to GetIntPoint made before SetIntPoint will cause a segfault which is not ideal but at least it halts execution which is probably for the best.  If we had a default integration point then execution would continue but with invalid data which would be bad (and difficult to debug).

I think the better alternative is to add a check for the NULL pointer which developers could use inside MFEM_ASSERT or MFEM_VERIFY to issue warnings in debug mode.  Then we would know that we have forgotten to a set proper integration point.

**Comment by YohannDudouit** [comment]
@mlstowell I guess this is the reason behind RAII, I think that we should not construct object if the constructor does not build an object that is ready to use (This is called a broken object). The construction of the object should be delayed until the point where you have all the information needed to construct a well formed object. I agree that there is cases where it is not possible, I don't think this is the case for an `ElementTransformation`.
Using the `explicit` keyword for the default constructor of `IntegratorRule` could solve some of the issues you're mentioning (or deleting it, _i.e._ making it `private`? since I'm not sure a default `IntegrationRule` has meaning). I agree that using the default constructor does not solve this problem. I had a similar problem recently with the `IntegrationRule` contained in `NonlinearFormIntegrator` who was `NULL`. I think we should be more careful with the use of this kind of function that can return `NULL`, because we don't always have an easy way to figure out what the value should be (because it should **not** be `NULL`). The problem is that programmers are lazy by essence, they write only what is necessary for a code to work, so if we let them set to `NULL` something that they don't use, it's likely that they do it, and then a user of this same code does not understand why something that shouldn't be `NULL` is `NULL`... This is why `NULL` makes poor API.

The thing that I don't like with the check for `NULL` is that it does not solve our problem, because we want the value that this `NULL` should have had, knowing that this is `NULL` barely helps in reconstructing the value. The best way to solve this problem, in my opinion, would be to forbid to set it to `NULL` and see where the code does not compile, and set the `IntegrationRule` at this point. I agree that it might result in several compilation errors which might not be trivial to solve.

How where you thinking of recovering the value of this `IntegrationRule` or `IntegrationPoint`?

**Comment by YohannDudouit** [comment]
@mlstowell  Looking closer at `ElementTransformation` it looks like most of its attributes are here to do some kind of caching. Usually, this kind of attributes is marked with the `mutable` keyword, it would also be good for the `const` correctness of mfem, because then most methods of `ElementTransformation` class could be `const`. I think it would make sense since evaluating the transformation, Jacobian, etc... does (should?) not conceptually change the state of an `ElementTransformation`.
```
class ElementTransformation
{
private:
   mutable const IntegrationPoint *IntPoint;
   mutable DenseMatrix dFdx, adjJ, invJ;
   mutable double Wght;
   mutable int EvalState;
   enum StateMasks
   {
      JACOBIAN_MASK = 1,
      WEIGHT_MASK   = 2,
      ADJUGATE_MASK = 4,
      INVERSE_MASK  = 8
   };
   int geom, space_dim;

   // Evaluate the Jacobian of the transformation at the IntPoint and store it
   // in dFdx.
   virtual const DenseMatrix &EvalJacobian() const = 0;

   double EvalWeight() const;
   const DenseMatrix &EvalAdjugateJ() const;
   const DenseMatrix &EvalInverseJ() const;

public:
   int Attribute, ElementNo;//really? mutable too?

   ElementTransformation()
   : IntPoint(new IntegrationPoint)//memory leak? we don't have to store a pointer... copying 4 values is not really expensive
   , EvalState(0)
   , Attribute(-1)
   , ElementNo(-1)
   {}

   void SetIntPoint(const IntegrationPoint *ip) const //Why a pointer here and a reference bellow? change this one to a reference? Since we don't want to evaluate at `NULL`... or is it a feature?
   { IntPoint = ip; EvalState = 0; }
   const IntegrationPoint &GetIntPoint() const { return *IntPoint; }

   virtual void Transform(const IntegrationPoint &, Vector &) const = 0;
   virtual void Transform(const IntegrationRule &, DenseMatrix &) const = 0;

   /// Transform columns of 'matrix', store result in 'result'.
   virtual void Transform(const DenseMatrix &matrix, DenseMatrix &result) const = 0;

   /** @brief Return the Jacobian matrix of the transformation at the currently
       set IntegrationPoint, using the method SetIntPoint(). */
   /** The dimensions of the Jacobian matrix are physical-space-dim by
       reference-space-dim. The first column contains the x derivatives of the
       transformation, the second -- the y derivatives, etc. */
   const DenseMatrix &Jacobian() const
   { return (EvalState & JACOBIAN_MASK) ? dFdx : EvalJacobian(); }

   double Weight() const { return (EvalState & WEIGHT_MASK) ? Wght : EvalWeight(); }

   const DenseMatrix &AdjugateJacobian() const
   { return (EvalState & ADJUGATE_MASK) ? adjJ : EvalAdjugateJ(); }

   const DenseMatrix &InverseJacobian() const
   { return (EvalState & INVERSE_MASK) ? invJ : EvalInverseJ(); }

   virtual int Order() const = 0;
   virtual int OrderJ() const = 0;
   virtual int OrderW() const = 0;
   /// Order of adj(J)^t.grad(fi)
   virtual int OrderGrad(const FiniteElement *fe) const = 0;

   /// Return the Geometry::Type of the reference element.
   int GetGeometryType() const { return geom; }

   /// Return the dimension of the reference element.
   int GetDimension() const { return Geometry::Dimension[geom]; }

   /// Get the dimension of the target (physical) space.
   /** We support 2D meshes embedded in 3D; in this case the function will
       return "3". */
   int GetSpaceDim() const { return space_dim; }

   /** @brief Transform a point @a pt from physical space to a point @a ip in
       reference space. */
   /** Attempt to find the IntegrationPoint that is transformed into the given
       point in physical space. If the inversion fails a non-zero value is
       returned. This method is not 100 percent reliable for non-linear
       transformations. */
   virtual int TransformBack(const Vector &pt, IntegrationPoint &ip) const = 0;

   virtual ~ElementTransformation() { }
};
```
This does not directly solves your problem, but we could propose a cleaner API that goes in the direction I proposed. Then, it would solve your problem I think.

**Comment by YohannDudouit** [comment]
I think that having:
```
mutable IntegrationPoint IntPoint;
```
would solve your problem. I think this is reasonable, because storing the pointer is not really less expensive, because when we use a new point we need to read its values, thus copying the new `IntegrationPoint` should be relatively inexpensive.

**Comment by YohannDudouit** [comment]
I think that the idea of `GetIntPoint()` is to give you the point at which `DenseMatrix dFdx, adjJ, invJ; double Wght;` are cached so that you can avoid a potentially expensive computation of these. You shouldn't be worried to use the method `SetIntPoint()` as you need it.

**Comment by mlstowell** [comment]
This issue will be resolved by PR #593
